### PR TITLE
optimize graph mode output, use proper scope for lib.pipe, fix cpe auto-extension normalization

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/lib/ast.py [hunk 1]
```diff
@@ -511,6 +510,7 @@ async def todogenr():
                 if doedges:
                     async for verb, n2iden in node.iterEdgesN1():
                         edges.append((n2iden, {'verb': verb}))
+                        await asyncio.sleep(0)
```
Comment (vEpiphyte) [inline synapse/lib/ast.py:513]: This is definitely a hot loop; but should the Node.iterEdgesN1() and Node.iterEdgesN2 APIs yield instead?

Comment (invisig0th) [reply synapse/lib/ast.py:513]: Might be an ever better optimization to mergegenr the edges from the layers rather than using a spooled set...

Comment (invisig0th) [reply synapse/lib/ast.py:513]: ( i'd like to keep this sleep(0) higher in the call stack so migrations etc can zoom

**Review by mikemoritz — APPROVED** [review]

File: synapse/models/infotech.py [hunk 2]
```diff
@@ -112,9 +112,8 @@ def _normPyStr(self, valu):
         text = valu.lower()
         if text.startswith('cpe:2.3:'):
             parts = cpesplit(text[8:])
-            if len(parts) != 11:
-                mesg = f'CPE 2.3 string has {len(parts)} fields, expected 11.'
-                raise s_exc.BadTypeValu(valu=valu, mesg=mesg)
+            extsize = 11 - len(parts)
+            parts.extend(['*' for _ in range(extsize)])
```
Comment (vEpiphyte) [inline synapse/models/infotech.py:116]: This allows invalid CPE 2.3 strings through.

```
cli> storm [it:sec:cpe=cpe:2.3:a:vendor:product:version:update:edition:lng:sw_edition:target_sw:target_hw:other:invald]
Executing query at 2023/06/16 16:46:57.052
..............
it:sec:cpe=cpe:2.3:a:vendor:product:version:update:edition:lng:sw_edition:target_sw:target_hw:other:invald
        .created = 2023/06/16 16:46:57.079
        :edition = edition
        :language = lng
        :other = other
        :part = a
        :product = product
        :sw_edition = sw_edition
        :target_hw = target_hw
        :target_sw = target_sw
        :update = update
        :v2_2 = cpe:/a:vendor:product:version:update:edition:lng
        :vendor = vendor
        :version = version
complete. 1 nodes in 29 ms (34/sec).
```

File: synapse/tests/test_model_infotech.py [hunk 3]
```diff
@@ -1284,9 +1281,6 @@ async def test_infotech_cpes(self):
             with self.raises(s_exc.BadTypeValu):
                 cpe23.norm('cpe:/a:vertex:synapse:0:1:2:3:4:5:6:7:8:9')
 
-            with self.raises(s_exc.BadTypeValu):
```
Comment (vEpiphyte) [inline synapse/tests/test_model_infotech.py:1287]: This should have a test showing the extended value

**Review by vEpiphyte — CHANGES_REQUESTED** [review]
Don't allow invalid cpe 2.3 strings through norm

**Review by vEpiphyte — APPROVED** [review]

**Pull request merged**